### PR TITLE
Avoid invalid array accesses to partitionFileArray

### LINKED ISSUE #2828: backend crash by user command
During testing I ran into this query that will crash the backend by any user.

```
postgres=# SELECT worker_hash_partition_table(42,1,'SELECT a FROM generate_series(1,100) AS a', 'a', 23, ARRAY[0]);
server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
Time: 8.579 ms
!>
```

Crux is probably in the length of the array.
Low priority

### PR DESCRIPTION
**Pull request by serprex**
DESCRIPTION: Prevent segfault in worker_partition_protocol edgecase

Closes #2828

### CODE REVIEW

File: src/backend/distributed/worker/worker_partition_protocol.c [hunk 1]
```diff
@@ -942,6 +942,12 @@ FilterAndPartitionTable(const char *filterQuery,
 	while (SPI_processed > 0)
 	{
 		int rowIndex = 0;
+
+		if (fileCount == 0)
```
Comment (serprex) [inline src/backend/distributed/worker/worker_partition_protocol.c]: Deferred error until here so that we don't mind a lack of shards if there's no results

**Comment by serprex** [comment]
The test failure is caused because now when `intermediate_results` test runs before `multi_query_directory_cleanup` test, there's a left over job directory from having an error raised

I lean towards adding a test function to clear PG_JOB_CACHE_DIR

**Comment by thanodnl** [comment]
Change all in all looks good, but I guess the test failures are due to this change (possibly by the added test).

They are super unstable now, too unstable to merge. If you think your described function is easy to implement lets go ahead with that.

**Comment by serprex** [comment]
@thanodnl implemented. Not sure if it's considered a bug that job directories are getting left around when errors are raised, I think this is something that happens when calling `worker_hash_partition_table` directly, whereas usually the caller cleans up afterwards. But that isn't a regression introduced here, since a) stray directories are better than segfaults & b) some other inputs could cause errors to be raised already

**Pull request merged**